### PR TITLE
Add LTX-2.X IC LoRA and HDR Pipelines

### PR DESCRIPTION
**Pull request by dg845**
# What does this PR do?

This PR adds two new LTX-2.X pipelines: `LTX2InContextPipeline`, which supports in-context (IC) conditioning (used for example by some IC LoRAs) and `LTX2HDRPipeline`, which supports the newly released high dynamic range (HDR) pipeline (and [HDR IC-LoRA](https://huggingface.co/Lightricks/LTX-2.3-22b-IC-LoRA-HDR)) introduced in the [LTX LumiVid paper](https://huggingface.co/papers/2604.11788).

This PR also updates `LTX2ConditionPipeline` to follow the [LTX-2 repo](https://github.com/Lightricks/LTX-2)'s current image conditioning strategy, which overwrites the noisy latents only for first-frame (I2V) conditions and treats non-first-frame as keyframe conditions which are appended to the noisy latents. (Previously, the pipeline overwrote the noisy latents at all latent indices.)

Here is an example of using `LTX2InContextPipeline` with the [`Lightricks/LTX-2-19b-LoRA-Camera-Control-Dolly-In`](https://huggingface.co/Lightricks/LTX-2-19b-LoRA-Camera-Control-Dolly-In) IC LoRA, which biases the generated video toward panning in on the subject:

<details>
    <summary>LTX-2.3 IC LoRA Example Script</summary>

```python
import torch

from diffusers import LTX2InContextPipeline
from diffusers.pipelines.ltx2.export_utils import encode_video
from diffusers.pipelines.ltx2.utils import DEFAULT_NEGATIVE_PROMPT


pipe = LTX2InContextPipeline.from_pretrained("dg845/LTX-2.3-Diffusers", torch_dtype=torch.bfloat16)
pipe.enable_model_cpu_offload(device="cuda:0")
pipe.load_lora_weights(
    "Lightricks/LTX-2-19b-LoRA-Camera-Control-Dolly-In",
    adapter_name="ic_lora",
    weight_name="ltx-2-19b-lora-camera-control-dolly-in.safetensors",
)
pipe.set_adapters("ic_lora", 1.0)

# If the IC LoRA uses reference conditions, you can specify them as follows:
# reference_video = load_video("reference.mp4")
# ref_cond = LTX2ReferenceCondition(frames=reference_video, strength=1.0)

prompt = "A flowing river in a forest"
frame_rate = 24.0
video, audio = pipe(
    prompt=prompt,
    negative_prompt=DEFAULT_NEGATIVE_PROMPT,
    # reference_conditions=[ref_cond],
    width=768,
    height=512,
    num_frames=121,
    frame_rate=frame_rate,
    num_inference_steps=30,
    guidance_scale=3.0,
    output_type="np",
    return_dict=False,
)

encode_video(
    video[0],
    fps=frame_rate,
    audio=audio[0].float().cpu(),
    audio_sample_rate=pipe.vocoder.config.output_sampling_rate,
    output_path="ltx2_ic_lora_output.mp4",
)
```
</details>

https://github.com/user-attachments/assets/6e487469-40ee-40fc-be3b-88b4eacbed98

And here is an example of using `LTX2HDRPipeline` with the [`Lightricks/LTX-2.3-22b-IC-LoRA-HDR`](https://huggingface.co/Lightricks/LTX-2.3-22b-IC-LoRA-HDR) HDR IC-LoRA, using the video above as the reference:

<details>
    <summary>LTX-2.3 HDR IC LoRA Example Script</summary>

```python
import torch
from safetensors import safe_open

from diffusers import LTX2HDRPipeline
from diffusers.pipelines.ltx2.export_utils import encode_hdr_tensor_to_mp4
from diffusers.pipelines.ltx2.pipeline_ltx2_hdr_lora import LTX2HDRReferenceCondition
from diffusers.pipelines.ltx2.utils import DISTILLED_SIGMA_VALUES
from diffusers.utils import load_video


pipe = LTX2HDRPipeline.from_pretrained("dg845/LTX-2.3-Distilled-Diffusers", torch_dtype=torch.bfloat16)
pipe.enable_model_cpu_offload(device="cuda")
pipe.load_lora_weights(
    "Lightricks/LTX-2.3-22b-IC-LoRA-HDR",
    adapter_name="hdr_lora",
    weight_name="ltx-2.3-22b-ic-lora-hdr-0.9.safetensors",
)
pipe.set_adapters("hdr_lora", 1.0)

reference_video = load_video("ltx2_ic_lora_output.mp4")
ref_cond = LTX2HDRReferenceCondition(frames=reference_video, strength=1.0)

# Load pre-computed HDR LoRA connector embeddings.
with safe_open(
    "/path/to/ltx-2.3-22b-ic-lora-hdr-scene-emb.safetensors", framework="pt", device="cuda"
) as f:
    connector_video_embeds = f.get_tensor("video_context")
    connector_audio_embeds = f.get_tensor("audio_context")

# `hdr_video` is a linear HDR tensor of shape (batch, frames, H, W, C).
hdr_video = pipe(
    reference_conditions=[ref_cond],
    connector_video_embeds=connector_video_embeds,
    connector_audio_embeds=connector_audio_embeds,
    width=768,
    height=512,
    num_frames=121,
    frame_rate=24.0,
    num_inference_steps=8,
    sigmas=DISTILLED_SIGMA_VALUES,
    guidance_scale=1.0,
    output_type="pt",
    return_dict=False,
)[0]

# Convert the HDR video to a SDR sRGB-tonemapped `.mp4` video.
# You can also save the output to EXR using `save_hdr_video_frames_as_exr`.
# A custom tone-mapper can be specified via the `tone_mapping_fn` argument.
encode_hdr_tensor_to_mp4(
    hdr_video[0],
    output_mp4="ltx2_hdr_lora_output.mp4",
    frame_rate=24.0,
)
```
</details>

https://github.com/user-attachments/assets/b397071d-4107-4bcf-990a-d7b3cdaab584

This uses the default tone-mapper that simply clips HDR values to `[0, 1]`, which is also used by the original LTX-2 code.

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu
@sayakpaul
@linoytsaban

### CODE REVIEW

File: src/diffusers/pipelines/ltx2/export_utils.py [hunk 1]
```diff
@@ -189,3 +191,274 @@ def encode_video(
         _write_audio(container, audio_stream, audio, audio_sample_rate)
 
     container.close()
+
+
+# Adapted from ltx_pipelines.utils.media_io.save_exr_tensor
+# https://github.com/Lightricks/LTX-2/blob/41d924371612b692c0fd1e4d9d94c3dfb3c02cb3/packages/ltx-pipelines/src/ltx_pipelines/utils/media_io.py#L609
+def save_exr_tensor(
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/export_utils.py]: I think  with this PR, `encode_hdr_tensor_to_mp4` is enough, no?
can we remove all the other functions here?

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py [hunk 2]
```diff
@@ -1005,18 +1168,18 @@ def __call__(
         width: int = 768,
         num_frames: int = 121,
         frame_rate: float = 24.0,
-        num_inference_steps: int = 40,
+        num_inference_steps: int = 30,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py]: did they change the default since the release? :exploding_head:

Comment (dg845) [reply src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py]: I ended up changing the `LTX2ConditionPipeline.__call__` defaults to the LTX-2.3 defaults (since the implementation now follows the newest behavior rather than the original LTX-2.0 behavior); `num_inference_steps=30` is the default for LTX-2.3, while `num_inference_steps=40` is the default for LTX-2.0.

Comment (yiyixuxu) [reply src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py]: Let's treat the update on image conditioning strategy as a bug fix and not change anything else

Comment (dg845) [reply src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py]: Changed the params back to the previous default values.

File: src/diffusers/pipelines/ltx2/export_utils.py [hunk 3]
```diff
@@ -189,3 +191,89 @@ def encode_video(
         _write_audio(container, audio_stream, audio, audio_sample_rate)
 
     container.close()
+
+
+def simple_tone_map(x: np.ndarray) -> np.ndarray:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/export_utils.py]: let's just move this line of code inside `encode_hdr_tensor_to_mp4`

File: src/diffusers/pipelines/ltx2/export_utils.py [hunk 4]
```diff
@@ -189,3 +191,89 @@ def encode_video(
...
+    Applies a very simple tone-mapping function on (scene-referred) linear light which simply clips values above `1.0`
+    to `1.0`. This is what the original LTX-2.X code does, but you probably want to do some non-trivial tone-mapping to
+    make the sample look better.
+    """
+    return np.clip(x, 0.0, 1.0)
+
+
+# Adapted from ltx_pipelines.utils.medio_io._linear_to_srgb
+# https://github.com/Lightricks/LTX-2/blob/41d924371612b692c0fd1e4d9d94c3dfb3c02cb3/packages/ltx-pipelines/src/ltx_pipelines/utils/media_io.py#L644
+def linear_to_srgb(x: np.ndarray) -> np.ndarray:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/export_utils.py]: same here, let's move this small function

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py [hunk 5]
```diff
@@ -1028,7 +1191,7 @@ def __call__(
         negative_prompt_attention_mask: torch.Tensor | None = None,
         decode_timestep: float | list[float] = 0.0,
         decode_noise_scale: float | list[float] | None = None,
-        use_cross_timestep: bool = False,
+        use_cross_timestep: bool = True,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_condition.py]: why do we change this?

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 6]
```diff
@@ -0,0 +1,1601 @@
...
+                negative_prompt_attention_mask=negative_prompt_attention_mask,
+                max_sequence_length=max_sequence_length,
+                device=device,
+            )
+            if self.do_classifier_free_guidance:
+                prompt_embeds = torch.cat([negative_prompt_embeds, prompt_embeds], dim=0)
+                prompt_attention_mask = torch.cat([negative_prompt_attention_mask, prompt_attention_mask], dim=0)
+
+            tokenizer_padding_side = "left"
+            if getattr(self, "tokenizer", None) is not None:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]: can we move this to `__init__`?

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 7]
```diff
@@ -0,0 +1,1601 @@
...
+
+        # 5. Prepare audio latents. Audio is discarded at the end, but the transformer's audio branch still runs so
+        # we need well-formed audio inputs. Audio guidance is fixed so no extra audio-only forward passes fire.
+        duration_s = num_frames / frame_rate
+        audio_latents_per_second = (
+            self.audio_sampling_rate / self.audio_hop_length / float(self.audio_vae_temporal_compression_ratio)
+        )
+        audio_num_frames = round(duration_s * audio_latents_per_second)
+
+        num_mel_bins = self.audio_vae.config.mel_bins if getattr(self, "audio_vae", None) is not None else 64
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]: can you move this to __init__

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 8]
```diff
@@ -0,0 +1,1601 @@
...
+        latent_mel_bins = num_mel_bins // self.audio_vae_mel_compression_ratio
+        num_channels_latents_audio = (
+            self.audio_vae.config.latent_channels if getattr(self, "audio_vae", None) is not None else 8
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]: same here too

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 9]
```diff
@@ -0,0 +1,1601 @@
...
+        # 6. Prepare timesteps
+        sigmas = np.linspace(1.0, 1 / num_inference_steps, num_inference_steps) if sigmas is None else sigmas
+        mu = calculate_shift(
+            self.scheduler.config.get("max_image_seq_len", 4096),
+            self.scheduler.config.get("base_image_seq_len", 1024),
+            self.scheduler.config.get("max_image_seq_len", 4096),
+            self.scheduler.config.get("base_shift", 0.95),
+            self.scheduler.config.get("max_shift", 2.05),
+        )
+        audio_scheduler = copy.deepcopy(self.scheduler)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]: maybe adding another scheduler component? audio_sehcduler?

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 10]
```diff
@@ -0,0 +1,1601 @@
...
+                    prompt_embeds = callback_outputs.pop("prompt_embeds", prompt_embeds)
+
+                if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
+                    progress_bar.update()
+
+                if XLA_AVAILABLE:
+                    xm.mark_step()
+
+        # Silence unused-variable lints for `audio_latent_model_input` / `latent_mel_bins`.
+        del audio_latent_model_input, latent_mel_bins
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]:
```suggestion
```

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 11]
```diff
@@ -0,0 +1,1601 @@
...
+            # VAE decode returns a video tensor in the VAE's native range ([-1, 1]).
+            decoded = self.vae.decode(latents, timestep, return_dict=False)[0]
+            # HDR postprocess: LogC3 decompress → linear HDR [0, ∞). Always float32 for HDR fidelity.
+            video = self.hdr_video_processor.postprocess_hdr_video(decoded)
+
+            # Format output (batch, frames, H, W, channels).
+            video = video.permute(0, 2, 3, 4, 1).contiguous()
+            if output_type == "np":
+                video = video.cpu().numpy()
+            elif output_type != "pt":
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]: move this inot check_inputs?

File: src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py [hunk 12]
```diff
@@ -0,0 +1,1601 @@
...
+
+            latents = self._denormalize_latents(
+                latents, self.vae.latents_mean, self.vae.latents_std, self.vae.config.scaling_factor
+            )
+            latents = latents.to(self.vae.dtype)
+
+            # VAE decode returns a video tensor in the VAE's native range ([-1, 1]).
+            decoded = self.vae.decode(latents, timestep, return_dict=False)[0]
+            # HDR postprocess: LogC3 decompress → linear HDR [0, ∞). Always float32 for HDR fidelity.
+            video = self.hdr_video_processor.postprocess_hdr_video(decoded)
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/ltx2/pipeline_ltx2_hdr_lora.py]: we already have a custom image processor, should we just make it take output_type and return expected output?

**Review by yiyixuxu — COMMENTED** [review]
left a bit more feedbacks, 
can we add tests too?

**Review by yiyixuxu — APPROVED** [review]
thanks!
